peer.go

```diff
@@ -71,3 +95,35 @@ func (peer enetPeer) SendPacket(packet Packet, channel uint8) error {
 	)
 	return nil
 }
+
+func (peer enetPeer) SetData(data []byte) {
+	if len(data) > MaxPeerDataLength {
+		panic(fmt.Sprintf("cannot store data with len > %d", MaxPeerDataLength))
+	}
+
+	if data == nil {
+		peer.cPeer.data = nil
+		return
+	}
+
+	// First byte is how long our slice is.
+	b := make([]byte, len(data)+1)
+	b[0] = byte(len(data))
+	copy(b[1:], data)
+	peer.cPeer.data = unsafe.Pointer(&b[0])
```

Comment:
How long does the data stay valid for? I assume there will come a point where `b` gets freed by Go. Does the data need to be stored in the `enetPeer` object for the unsafe pointer to remain valid?

I know that in the tests you check for validity after a GC call, but is that guaranteed? Could it still point to memory that is "valid data" but still freed memory?

Reply:
Additionally, this limits the data length to 255 bytes, no?

Reply:
My understanding is that because Go created both the byte slice and unsafe pointer, GC won't touch it for as long as the pointer lives (from some reading [here](https://groups.google.com/g/golang-nuts/c/yNis7bQG_rY/m/yaJFoSx1hgIJ)). My thinking was that putting this in the C struct (`peer.cPeer.data`) would keep the pointer alive as far as Go's GC is concerned, but now I'm not sure. This is why I covered it in the test, but yes you're quite right that maybe this is passing accidentally. I'll need to do more reading to comment more confidently.

Reply:
> How long does the data stay valid for? I assume there will come a point where b gets freed by Go. Does the data need to be stored in the enetPeer object for the unsafe pointer to remain valid?

I've addressed this in my latest change and explained my thinking in [the commit message](https://github.com/vaeryn-uk/go-enet/commit/951bfb7266d6a7f1e27a98aa3a3a629149533bf4).

Reading https://pkg.go.dev/cmd/cgo, using `C.Cbytes` ensures this data will stay valid in C's heap and out of reach of Go's GC. As mentioned in the commit, I couldn't find a neat way to auto-free this, so unfortunately the best I can think of is having the user do it.

> I know that in the tests you check for validity after a GC call, but is that guaranteed? Could it still point to memory that is "valid data" but still freed memory?

You were absolutely right, it was not guaranteed which `cgocheck` highlighted.

> Additionally, this limits the data length to 255 bytes, no?

I've increased this so the limit should be sufficient for any practical use-case.